spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -0,0 +1,371 @@
+   * @param fromType         The original type of the column 'value' in the test table.
+   * @param toType           The type to use when changing the type of column 'value'
+   * @param initialValues    The initial values to insert in column 'value after table creation,
+   *                         using type `fromType`
+   * @param additionalValues Additional values to insert after changing the type of the column
+   *                         'value' to `toType`.
+   */
+  case class TypeEvolutionTestCase(
+      fromType: DataType,
+      toType: DataType,
+      initialValues: Seq[String],
+      additionalValues: Seq[String] = Seq.empty) {
+    def initialValuesDF: DataFrame =
+      initialValues.toDF("value").select($"value".cast(fromType))
+
+    def additionalValuesDF: DataFrame =
+      additionalValues.toDF("value").select($"value".cast(toType))
+
+    def expectedResult: DataFrame =
+      initialValuesDF.union(additionalValuesDF).select($"value".cast(toType))
+  }
+
+  // Type changes that are supported by all Parquet readers. Byte, Short, Int are all stored as
+  // INT32 in parquet so these changes are guaranteed to be supported.
+  private val supportedTestCases: Seq[TypeEvolutionTestCase] = Seq(
+    TypeEvolutionTestCase(ByteType, ShortType,
+      Seq("1", "2", Byte.MinValue.toString),
+      Seq("4", "5", Int.MaxValue.toString)),
+    TypeEvolutionTestCase(ByteType, IntegerType,
+      Seq("1", "2", Byte.MinValue.toString),
+      Seq("4", "5", Int.MaxValue.toString)),
+    TypeEvolutionTestCase(ShortType, IntegerType,
+      Seq("1", "2", Byte.MinValue.toString),
+      Seq("4", "5", Int.MaxValue.toString))
+  )
+
+  for {
+    testCase <- supportedTestCases
+    partitioned <- BOOLEAN_DOMAIN
+  } {
+    test(s"type widening ${testCase.fromType.sql} -> ${testCase.toType.sql}, " +
+      s"partitioned=$partitioned") {
+      def writeData(df: DataFrame): Unit = if (partitioned) {
+        // The table needs to have at least 1 non-partition column, use a dummy one.
+        append(df.withColumn("dummy", lit(1)), partitionBy = Seq("value"))
+      } else {
+        append(df)
+      }
+
+      writeData(testCase.initialValuesDF)
+      sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN value TYPE ${testCase.toType.sql}")
+      withAllParquetReaders {
+        assert(readDeltaTable(tempPath).schema("value").dataType === testCase.toType)
+        checkAnswer(readDeltaTable(tempPath).select("value"),
+          testCase.initialValuesDF.select($"value".cast(testCase.toType)))
+      }
+      writeData(testCase.additionalValuesDF)
+      withAllParquetReaders {
+        checkAnswer(readDeltaTable(tempPath).select("value"), testCase.expectedResult)
+      }
+    }
+  }
+
+  // Test type changes that aren't supported.
+  private val unsupportedNonTestCases: Seq[TypeEvolutionTestCase] = Seq(
+    TypeEvolutionTestCase(IntegerType, ByteType,
+      Seq("1", "2", Int.MinValue.toString)),
```

Comment:
It looks like the optionality of the second sequence parameter in TypeEvolutionTestCase is only to support the negative case. It might make sense to turn this into a separate "negative case" class, because the optional parameter allowed a missing second sequence parameter above.

Reply:
Good idea, I split the test case classes in two